go/pkg/casng/batching.go

```diff
@@ -101,29 +101,33 @@ func (u *BatchingUploader) MissingBlobs(ctx context.Context, digests []digest.Di
 // The errors returned are either from the context, ErrGRPC, ErrIO, or ErrCompression. More errors may be wrapped inside.
 // If an error was returned, the returned stats may indicate that all the bytes were sent, but that does not guarantee that the server committed all of them.
 func (u *BatchingUploader) WriteBytes(ctx context.Context, name string, r io.Reader, size, offset int64) (Stats, error) {
+	startTime := time.Now()
 	if !u.streamThrottle.acquire(ctx) {
 		return Stats{}, ctx.Err()
 	}
 	defer u.streamThrottle.release()
+	log.V(3).Infof("[casng] upload.write_bytes.throttle.duration; start=%d, end=%d", startTime.UnixNano(), time.Now().UnixNano())
```

Comment:
optional: maybe calculate duration instead of providing start end end. Also: wouldn't end be already present in log's timestamp?

Reply:
The lifetime duration is the interesting value. To get it, the durations need to be merged into the longest non-overlapping interval to exclude redundant ranges when goroutines were running in parallel. Logging durations would truncate the information necessary to calculate that. The log timestamp is controlled by the logger which may be delayed.

This should ideally be a metric measurement rather than a log message, but that's for another day.